cmd/zc_traverser_blob.go

```diff
@@ -137,57 +138,73 @@ func (t *blobTraverser) traverse(preprocessor objectMorpher, processor objectPro
 	// as a performance optimization, get an extra prefix to do pre-filtering. It's typically the start portion of a blob name.
 	extraSearchPrefix := filterSet(filters).GetEnumerationPreFilter(t.recursive)
 
-	for marker := (azblob.Marker{}); marker.NotDone(); {
-
-		// see the TO DO in GetEnumerationPreFilter if/when we make this more directory-aware
+	// Define how to enumerate its contents
+	// This func must be thread safe/goroutine safe
+	enumerateOneDir := func(dir parallel.Directory, enqueueDir func(parallel.Directory), enqueueOutput func(parallel.DirectoryEntry, error)) error {
+		currentDirPath := dir.(string)
+		for marker := (azblob.Marker{}); marker.NotDone(); {
+			lResp, err := containerURL.ListBlobsHierarchySegment(t.ctx, marker, "/", azblob.ListBlobsSegmentOptions{Prefix: currentDirPath})
+			if err != nil {
+				return fmt.Errorf("cannot list files due to reason %s", err)
+			}
 
-		// look for all blobs that start with the prefix
-		// TODO optimize for the case where recursive is off
-		listBlob, err := containerURL.ListBlobsFlatSegment(t.ctx, marker,
-			azblob.ListBlobsSegmentOptions{Prefix: searchPrefix + extraSearchPrefix, Details: azblob.BlobListingDetails{Metadata: true}})
-		if err != nil {
-			return fmt.Errorf("cannot list blobs. Failed with error %s", err.Error())
-		}
+			// process the blobs returned in this result segment
+			for _, blobInfo := range lResp.Segment.BlobItems {
+				// if the blob represents a hdi folder, then skip it
+				if util.doesBlobRepresentAFolder(blobInfo.Metadata) {
+					continue
+				}
+
+				relativePath := strings.TrimPrefix(blobInfo.Name, searchPrefix)
+				adapter := blobPropertiesAdapter{blobInfo.Properties}
+				storedObject := newStoredObject(
+					preprocessor,
+					getObjectNameOnly(blobInfo.Name),
+					relativePath,
+					common.EEntityType.File(),
+					blobInfo.Properties.LastModified,
+					*blobInfo.Properties.ContentLength,
+					adapter,
+					adapter, // adapter satisfies both interfaces
+					common.FromAzBlobMetadataToCommonMetadata(blobInfo.Metadata),
+					blobUrlParts.ContainerName,
+				)
+				enqueueOutput(storedObject, nil)
+			}
 
-		// process the blobs returned in this result segment
-		for _, blobInfo := range listBlob.Segment.BlobItems {
-			// if the blob represents a hdi folder, then skip it
-			if util.doesBlobRepresentAFolder(blobInfo.Metadata) {
-				continue
+			// queue up the sub virtual directories if recursive is true
```

Comment:
That's true, but what about the case when there are too many queued directories? When I was reading TreeCrawler.go, I got the impression that in that case, we want the routines to process blobs first, before queuing up more directories, so I'm a bit confused as to what should be the right order.

Reply:
I think it's pretty hard to get too many. The code currently defines "too many" as 1 million in the queue.  Which won't happen often, if at all. (I _think_).  

If you can't enqueue the directories first, I'd be tempted to view that as a bug in the TreeCrawler, rather than in the code in this PR.

Also, I should note that the limits in TreeCrawler, and the way it responds to them, are just things I made up. I haven't done realistic testing (at the million-queued dirs scale) to verify that that's the right limit and that the way it responds to it is optimal.  But, even if it's not optimal, we know it won't _break_. It might just slow down. (I have tested that, by setting a smaller limit than 1 million!)

Reply:
It's also worth noting that, once it detects that its over 10% of the way to the limit, it switches inside itself to a depth-first strategy, which will tend to limit further growth in many if not most cases.